_posts/2021-02-03-#20721.md

```diff
@@ -0,0 +1,111 @@
+---
+layout: pr
+date: 2021-02-03
+title: "Move ping data to net_processing"
+pr: 20721
+authors: [jnewbery]
+components: ["p2p"]
+host: jnewbery
+status: upcoming
+commit:
+---
+
+_Notes and questions to follow soon!_
+
+## Background
+
+- We can think of the Bitcoin Core node as a stack with three levels:
+
+  - **Net** - keeps track of potential nodes to peer with, handles active
+    connections to nodes, and reads and writes messages to the wire. The main
+    class in the net layer is `CConnman`, which is defined in
+    [net.cpp](https://github.com/bitcoin/bitcoin/blob/16b784d9/src/net.h). Net
+    hands messages up the stack to ...
+
+  - **Net Processing** - reads and processes application-level messages, and
+    handles application-level view of peers. The main class in net
+    processing is `PeerManager`, which is declared in
+    [net_processing.h](https://github.com/bitcoin/bitcoin/blob/16b784d9/src/net_processing.h).
+    Net processing uses function calls like `ProcessNewBlock()` and
+    `AcceptToMemoryPool()` to call into ...
+
+  - **Validation** - is responsible for our view of the block chain and the UTXO
+    set. It may (or may not) include a mempool for unconfirmed transactions. There
+    are various classes and functions exposed to the rest of the node. [PR
+    20158](https://github.com/bitcoin/bitcoin/pull/20158) and linked PRs are an
+    attempt to clarify and rationlize the interface to validation.
+
+- In early versions of Bitcoin, this distinction wasn't as clear. The [first
+  commit](https://github.com/jnewbery/bitcoin/blob/4405b78d/main.cpp) in the
+  Bitcoin Core repository has a separate net.cpp file, but net processing and
+  validation are mixed in the main.cpp file. It was only in [PR
+  9260](https://github.com/bitcoin/bitcoin/pull/9260) in 2016 that net processing
+  and validation were split into separate files.
+
+- Even today, the division between net, net processing and validation is not as
+  clear as we'd like. Clarifying and rationalizing these interfaces is an
+  ongoing project. See, for example, [PR
+  15141](https://github.com/bitcoin/bitcoin/pull/15141), which moves the denial
+  of service logic from validation into net processing, [PR
+  20158](https://github.com/bitcoin/bitcoin/pull/20158), which rationalizes the
+  interface to validation, and [PR
+  19398](https://github.com/bitcoin/bitcoin/issues/19398), which moves
+  application-layer data and logic from net into net processing.
+
+- The high level goal of PR 19398 is to split responsibilities cleanly between
+  CConnman and PeerManager:
+
+  - C**Conn**man should be responsible for **connections** - accepting incoming
+    and opening outbound connections, opening and closing sockets, reading and
+    writing data on those sockets, and first-line-of-defence validation of
```

Comment:
I think it's "defense" unless it's a UK spelling I'm unaware of

Reply:
Indeed it is my good chap: https://dictionary.cambridge.org/dictionary/english/defence :gb:

Reply:
> Indeed it is my good chap: https://dictionary.cambridge.org/dictionary/english/defence :gb: 

TIL--thanks! 🇬🇧